34-cellular-modbus-client/firmware/src/main.cpp

```diff
@@ -0,0 +1,650 @@
+            const char errMsg[] = "No server_addr field in request.";
+            Serial.println(errMsg);
+            sendError(seqNum, errMsg);
+            ret = -1;
+        }
+        else if (!JIsPresent(body, "seq_num")) {
+            const char errMsg[] = "No seq_num field in request.";
+            Serial.println(errMsg);
+            sendError(seqNum, errMsg);
+            ret = -1;
+        }
+        else if (!JIsPresent(body, "func")) {
+            const char errMsg[] = "No func field in request.";
+            Serial.println(errMsg);
+            sendError(seqNum, errMsg);
+            ret = -1;
+        }
+        else if (!JIsPresent(body, "data")) {
+            const char errMsg[] = "No data field in request.";
+            Serial.println(errMsg);
+            sendError(seqNum, errMsg);
+            ret = -1;
+        }
+        else {
+            long int serverAddr = JGetInt(body, "server_addr");
+            long int funcCode = JGetInt(body, "func");
+            J *data = JGetObject(body, "data");
+
+            switch (funcCode) {
+                case READ_COILS:
+                case READ_DISCRETE_INPUTS:
+                    ret = readBits(serverAddr, funcCode, seqNum, data);
+                    break;
+                case READ_MULTIPLE_HOLDING_REGS:
+                case READ_INPUTS_REGS:
+                    ret = readRegisters(serverAddr, funcCode, seqNum, data);
+                    break;
+                case WRITE_SINGLE_COIL:
+                    ret = writeSingleCoil(serverAddr, seqNum, data);
+                    break;
+                case WRITE_SINGLE_REGISTER:
+                    ret = writeSingleRegister(serverAddr, seqNum, data);
+                    break;
+                case WRITE_MULTIPLE_COILS:
+                    ret = writeMultipleCoils(serverAddr, seqNum, data);
+                    break;
+                case WRITE_MULTIPLE_REGS:
+                    ret = writeMultipleRegisters(serverAddr, seqNum, data);
+                    break;
+                default:
+                    const char errMsg[] = "Function code not supported.";
+                    Serial.println(errMsg);
+                    sendError(seqNum, errMsg);
+                    ret = -1;
+                    break;
+            }
+        }
+    }
+
+    return ret;
+}
+
+void attnISR()
+{
+    // This flag will be read in the main loop. We keep the ISR lean and do all
+    // the logic in the main loop.
+    attnTriggered = true;
+}
+
+// Arm the ATTN interrupt.
+void attnArm()
+{
+    // Once ATTN has triggered, it stays set until explicitly reset. Reset it
+    // here. It will trigger again after a change to the watched Notefile.
+    attnTriggered = false;
+    J *req = notecard.newRequest("card.attn");
+    JAddStringToObject(req, "mode", "reset");
+    if (!notecard.sendRequest(req)) {
+        Serial.println("card.attn reset failed");
+    }
+}
+
+void setup()
+{
+    Serial.begin(115200);
+    while (!Serial);
```

Comment:
This is useful for development - you probably recall I left a similar thing in on of my PRs. :-)  Perhaps make it compile-time conditional? I was hoping platformIO added different symbols between debug and release builds but they don't - adding one manually gets messy in `platformio.ini` having to duplicate many of the build options like platform, framework, board etc.  in `[env:debug]`.

Thinking about typical developer use and CI/CD, making it an opt out feature could work well:

1. The initial experience is good for developers - the device waits until the serial connection so nothing is missed in the serial monitor.

2. CI/CD builds can easily define the opt out symbol as part of the release build. But having the wait for connect feature could be useful for testing to have serial output halted until connected.

3. For developer convenience when they want to build a deployable version, we add a commented-out build flag so the feature can be disabled just by removing the comment marker.

```
[env:bw_swan_r5]
build_flags = 
    	-D PIO_FRAMEWORK_ARDUINO_ENABLE_CDC # enable USB CDC (Serial)
	# -D NF34_NO_WAIT_FOR_SERIAL
```

What do you think?

Reply:
Sounds reasonable to me. I'll add a flag like that in.